### PR TITLE
Enable kv cache in training for eagle

### PR DESCRIPTION
**Pull request by yeyu-nvidia**

### CODE REVIEW

**Comment by chtruong814** [comment]
I think @yaoyu-33 and @ChenhanYu previously approved before this opened on Github.  @jaredcasper I think you had some feedback that was addressed before when this was a MR.  Please take another look.

**Comment by yeyu-nvidia** [comment]
@chtruong814 @jaredcasper any update on this PR's review?

**Review by ChenhanYu — APPROVED** [review]

**Review by pablo-garay — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]

**Review by ericharper — APPROVED** [review]

**Comment by yeyu-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/4ec84e09a46f45d6f01d315b1e77a2f4472b3d08

File: megatron/core/transformer/attention.py [hunk 1]
```diff
@@ -333,9 +333,15 @@ def _adjust_key_value_for_inference(
                     inference_context.key_value_memory_dict[self.layer_number]
                 )
 
-        if not inference_context.is_static_batching() or inference_context.sequence_len_offset > 0:
+        if (
+            not inference_context.is_static_batching() or inference_context.sequence_len_offset > 0
+        ) and (not self.training or not is_te_min_version("2.2.0")):
```
Comment (yobibyte) [inline megatron/core/transformer/attention.py:349]: Why do we need to check for self.training in the inference-side function?

Comment (yeyu-nvidia) [reply megatron/core/transformer/attention.py:349]: Because in ModelOpt speculative decoding training, we use kv cache (inference_context). So we have to check self.training to decide whether or not to set mask type to no_mask.

Comment (yeyu-nvidia) [reply megatron/core/transformer/attention.py:349]: We have added a note here to explain: https://github.com/NVIDIA/Megatron-LM/pull/1895/files#diff-cfcaf53b88d3893379f5522e1e2a6d0b15eba6158e964fb1892415de482aadf8R341

**Comment by yeyu-nvidia** [comment]
@ko3n1g @jaredcasper  Is there anything missing before we can merge this?

**Comment by yeyu-nvidia** [comment]
@BoxiangW @kvareddy @shanmugamr1992 can I get a review on this PR please?

**Comment by yeyu-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/35b0a010297657781697463985e3422600514029

**Comment by yeyu-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/214b96aca9e9add28684aa13e1db00a72796fba2

**Comment by yeyu-nvidia** [comment]
/ok to test [214b96a](https://github.com/NVIDIA/Megatron-LM/commit/214b96aca9e9add28684aa13e1db00a72796fba2)

**Comment by ko3n1g** [comment]
/ok to test [214b96a](https://github.com/NVIDIA/Megatron-LM/commit/214b96aca9e9add28684aa13e1db00a72796fba2)
